File: 0133-clone-graph/0133-clone-graph.py
"""
# Definition for a Node.
class Node(object):
    def __init__(self, val = 0, neighbors = None):
        self.val = val
        self.neighbors = neighbors if neighbors is not None else []
"""

class Solution(object):
    def __init__(self):
        self.hashmap = defaultdict(Node)
    def cloneGraph(self, node):
        """
        :type node: Node
        :rtype: Node
        """
        if node == None:
            return
        
        return self.DFSFunction(node)
        
    def DFSFunction(self, node):
        #1. Check all the necessary conditions
        if node in self.hashmap.keys():
            return self.hashmap[node]
        
        #2. Process the cell
        createdClone = Node(node.val, [])
        self.hashmap[node] = createdClone
            
        #3. Call the DFS function recursively 
        for neighbors in node.neighbors:
                createdClone.neighbors.append(self.DFSFunction(neighbors))
        
        return createdClone
        
        
        
    

File: 0133-clone-graph/0133-clone-graph.java
/*
// Definition for a Node.
class Node {
    public int val;
    public List<Node> neighbors;
    public Node() {
        val = 0;
        neighbors = new ArrayList<Node>();
    }
    public Node(int _val) {
        val = _val;
        neighbors = new ArrayList<Node>();
    }
    public Node(int _val, ArrayList<Node> _neighbors) {
        val = _val;
        neighbors = _neighbors;
    }
}
*/

class Solution {
    
    public Map<Node, Node> hashmap = new HashMap<>();
    
    //we start off with the Depth first search solution 
    public Node cloneGraph(Node node) {
        if(node == null){
            return null;
        }
        return this.DFSFunction(node);
    }
    
    public Node DFSFunction(Node node){
       //1. Check the necessary conditions 
        if(this.hashmap.containsKey(node)){
            return this.hashmap.get(node);
        }
        
        //2. Process the cell 
        Node clonedNode = new Node(node.val);
        this.hashmap.put(node,clonedNode );
        
        //3. Recusrive DFS Calls 
        for(Node neighbors: node.neighbors){
            clonedNode.neighbors.add(this.DFSFunction(neighbors));
            //this.hashmap.put(node, this.DFSFunction(neighbors));
        }
        
        return clonedNode;
    }
}